### PR TITLE
Wrap tar calls using fakeroot

### LINKED ISSUE #3523: Error extracting toolchain for armada38x
### Setup
_Package Name:_ ALL
_Package Version:_ ANY

_NAS Model:_ DS218j
_NAS Architecture:_ armada38x
_DSM version:_ 6.1/6.2

### Expected behavior
tmux/git/... should build using make arch-armada38x

### Actual behavior
Error due to bad extraction of toolchain

### Steps to reproduce
_1._ use the official docker dev container 
_2._ follow instructions to setup and build any package
_3._ try building git for instance

#### Other logs
_E.g. `command output`_
```
===>    Extracting for syno-armada38x
tar -xpf /spksrc/toolchains/syno-armada38x-6.1/../../distrib/toolchains/6.1/armada38x-gcc493_glibc220_hard-GPL.txz -C /spksrc/toolchains/syno-armada38x-6.1/work 
tar: arm-unknown-linux-gnueabi/share: Cannot mkdir: Permission denied
tar: arm-unknown-linux-gnueabi/share: Cannot mkdir: Permission denied
tar: arm-unknown-linux-gnueabi/share/locale: Cannot mkdir: No such file or directory
...
# find /spksrc/toolchains/syno-armada38x-6.1/work -ls
42733955    0 drwxr-xr-x   5 root     root          170 Nov 21 23:29 /spksrc/toolchains/syno-armada38x-6.1/work
42733960    0 -rw-r--r--   1 root     root            0 Nov 21 23:29 /spksrc/toolchains/syno-armada38x-6.1/work/.syno-armada38x-checksum_done
42733956    0 -rw-r--r--   1 root     root            0 Nov 21 23:28 /spksrc/toolchains/syno-armada38x-6.1/work/.syno-armada38x-download_done
42733961    0 dr-xr-xr-x   2 1000     1000           68 Jun 24  2016 /spksrc/toolchains/syno-armada38x-6.1/work/arm-unknown-linux-gnueabi
```

### LINKED ISSUE #3570: Build fails with Docker image on MacOS
_For new Package Requests, see the guidelines_  

### Setup
_Package Name:_ git
_Package Version:_

_NAS Model:_ ds418
_NAS Architecture:_ rtd1296
_DSM version:_

### Expected behavior
make arch-rtd1296
Make succeeds

### Steps to reproduce
_1._ Install docker.
_2._ Run docker environment
_3._ Git clone the tree
_4._ cd spk/git
_5._ make arch-rtd1296

#### Package log
_Check Package Center or `/usr/local/{package}/var/`_
```
root@f6dcda86815a:/spksrc/spksrc/spk/git# make arch-rtd1296
===>  Building package for arch rtd1296
make[1]: Entering directory '/spksrc/spksrc/spk/git'
===>  Set up toolchain 
===>    Extracting for syno-rtd1296
tar -xpf /spksrc/spksrc/toolchains/syno-rtd1296-6.1/../../distrib/toolchains/6.1/rtd1296-gcc494_glibc220_armv8-GPL.txz -C /spksrc/spksrc/toolchains/syno-rtd1296-6.1/work 
tar: aarch64-unknown-linux-gnueabi/build.log.bz2: Cannot open: Permission denied
tar: aarch64-unknown-linux-gnueabi/aarch64-unknown-linux-gnueabi: Cannot mkdir: Permission denied
tar: aarch64-unknown-linux-gnueabi/aarch64-unknown-linux-gnueabi: Cannot mkdir: Permission denied
...
```
#### Other logs
_E.g. `/var/log/messages` or `/var/log/synopkg.log`_
```
Insert log here
```

### LINKED ISSUE #3791: Following Developers How To fails building transmission using current docker image for arch-denverton-6.1
Title pretty much states the issue. 

Running docker ce 2.1.0.3 on mac 10.14.6 and fails extracting the tar file under distrib/tookchains. 

make arch-denverton-6.1
===>  Building package for arch denverton-6.1
make[1]: Entering directory '/spksrc/spk/transmission'
===>  Set up toolchain
===>    Extracting for syno-denverton
tar -xpf /spksrc/toolchains/syno-denverton-6.1/../../distrib/toolchains/6.1/denverton-gcc493_glibc220_linaro_x86_64-GPL.txz -C /spksrc/toolchains/syno-denverton-6.1/work
tar: x86_64-pc-linux-gnu/test-suite: Cannot mkdir: Permission denied
tar: x86_64-pc-linux-gnu/test-suite: Cannot mkdir: Permission denied
tar: x86_64-pc-linux-gnu/test-suite/gcc: Cannot mkdir: No such file or directory
tar: x86_64-pc-linux-gnu/test-suite: Cannot mkdir: Permission denied
tar: x86_64-pc-linux-gnu/test-suite/gcc/testsuite: Cannot mkdir: No such file or directory
tar: x86_64-pc-linux-gnu/test-suite: Cannot mkdir: Permission denied
tar: x86_64-pc-linux-gnu/test-suite/gcc/testsuite/objc: Cannot mkdir: No such file or directory
tar: x86_64-pc-linux-gnu/test-suite: Cannot mkdir: Permission denied
tar: x86_64-pc-linux-gnu/test-suite/gcc/testsuite/objc/execute: Cannot mkdir: No such file or directory
tar: x86_64-pc-linux-gnu/test-suite: Cannot mkdir: Permission denied
tar: x86_64-pc-linux-gnu/test-suite/gcc/testsuite/objc/execute/load.m: Cannot open: No such file or directory
tar: x86_64-pc-linux-gnu/test-suite: Cannot mkdir: Permission denied
tar: x86_64-pc-linux-gnu/test-suite/gcc/testsuite/objc/execute/function-message-1.m: Cannot open: No such file or directory
tar: x86_64-pc-linux-gnu/test-suite: Cannot mkdir: Permission denied
tar: x86_64-pc-linux-gnu/test-suite/gcc/testsuite/objc/execute/class_self-2.m: Cannot open: No such file or directory

### PR DESCRIPTION
**Pull request by ixs**
Some of the synology toolchain tarballs contain permissions that will make
extraction fail on MacOS Docker Host machines. Details are in #3570.

A simple fix for this issue is prepending the tar command with the
fakeroot command.

_Linked issues_: Fixes #3570, fixes #3523, closes #3524, #3428, fixes #3791

### CODE REVIEW

**Comment by hgy59** [comment]
Thanks for your contribution. This is a long stroy that we want to find a solution for.

We should not break working environments that do not have fakeroot installed.
So I probose to set `TAR_CMD ?= tar` as default.
Any user that wants to use the updated docker image with fakeroot installed can override the TAR_CMD like this

`docker run -it -v /home/whatever/spksrc:/spksrc -e TAR_CMD="fakeroot tar"  spksrc_image_with_fakeroot`

_I cannot verify whether fakeroot really solves the problem for MACOS as I use linux only as spksrc environment._

File: mk/spksrc.extract.mk [hunk 1]
```diff
@@ -7,16 +7,18 @@
 #  extract_target       (override with EXTRACT_TARGET)
 #  post_extract_target  (override with POST_EXTRACT_TARGET)
 EXTRACT_PATH ?= $(WORK_DIR)
+
 # Extract commands
+TAR_CMD = fakeroot tar
```
Comment (hgy59) [inline mk/spksrc.extract.mk:12]: see my comment
```suggestion
TAR_CMD ?= tar
```

**Comment by ixs** [comment]
> We should not break working environments that do not have fakeroot installed.
> So I propose to set `TAR_CMD ?= tar` as default.
> Any user that wants to use the updated docker image with fakeroot installed can override the TAR_CMD like this
>
> `docker run -it -v /home/whatever/spksrc:/spksrc -e TAR_CMD="fakeroot tar" spksrc_image_with_fakeroot`

Fair enough. Although I am not sure if that amount of backwards compatibility is a good idea in general. What if... hypothetically say a newer toolchain requires a not lzma but lzma2 to decompress something? That would require a new docker container as well. How are upgrades handled then?

That being said, how about dynamically testing wether fakeroot is needed or not? That would keep existing working containers using tar and only currently broken ones would need to be updated?
I was thinking of maybe using `ifeq` in the makefile to make the TAR_CMD assignment conditional, but I can't think of a way of making it work.

Alternatively we could just assign `TAR_CMD := $(shell {something})`.

A possible shell based check for the problem could look like this:
```
TDIR=$(mktemp -d "${EXTRACT_PATH}/tar.XXXXXX"); chmod 000 "${TDIR}"; touch "${TDIR}/test" 2> /dev/null && echo "tar" || echo "fakeroot tar"; chmod 700 "${TDIR}"; rm -rf "${TDIR}"
```

What do you think? Would something like that be more acceptable?

> _I cannot verify whether fakeroot really solves the problem for MACOS as I use linux only as spksrc environment._

Right. My test environment looked like this:

```
$ docker version
Client: Docker Engine - Community
 Cloud integration: 1.0.9
 Version:           20.10.5
 API version:       1.41
 Go version:        go1.13.15
 Git commit:        55c4c88
 Built:             Tue Mar  2 20:13:00 2021
 OS/Arch:           darwin/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.5
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.13.15
  Git commit:       363e9a8
  Built:            Tue Mar  2 20:15:47 2021
  OS/Arch:          linux/amd64
  Experimental:     true
 containerd:
  Version:          1.4.3
  GitCommit:        269548fa27e0089a8b8278fc4fc781d7f65a939b
 runc:
  Version:          1.0.0-rc92
  GitCommit:        ff819c7e9184c13b7c2607fe6c30ae19403a7aff
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
  ```

On that, I can confirm that "fakeroot tar" works and a `make all-archs` actually succeeds on MacOS, something that was not the case before.

**Review by ymartin59 — COMMENTED** [review]
Great idea!

**Pull request merged**